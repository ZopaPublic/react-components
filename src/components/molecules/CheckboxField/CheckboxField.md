### Summary

`<CheckboxField />` is a convenience wrapper around HTML `<input type="checkbox" />` and two **atoms**:

- [`<InputLabel />`](/#/Components/Atoms/InputLabel)
- [`<ErrorMessage />`](/#/Components/Atoms/ErrorMessage)

⚠️ &nbsp;Note that a `name` prop **must be provided** to automatically set:

- `htmlFor` prop in `<InputLabel />`
- `data-automation` prop in `<ErrorMessage />`
- `id` prop in `<input type="checkbox" />` for accessibility and automation purposes

### Examples

- Default

```tsx
import { CheckboxField } from '@zopauk/react-components';

<CheckboxField label="Do you love jazz?" name="check1" />;
```

- Hide control

```tsx
import { CheckboxField } from '@zopauk/react-components';

<CheckboxField label="Do you want to hide the icon?" name="hideControl" hideControl />;
```

- Default with multiline label

```tsx
import { RadioField } from '@zopauk/react-components';

<div style={{ maxWidth: '200px' }}>
  <CheckboxField label="This is an example of a multiline label" name="check2" />
</div>;
```

- With an error

```tsx
import { CheckboxField } from '@zopauk/react-components';

<CheckboxField label="Do you agree?" hasError={true} errorMessage="Oops! You forgot to check this" name="check3" />;
```

- Checked state

```tsx
import { CheckboxField } from '@zopauk/react-components';

<CheckboxField label="Do you love rap?" name="check4" defaultChecked={true} />;
```

- Checked and valid

```tsx
import { CheckboxField } from '@zopauk/react-components';

<CheckboxField label="Do you love rap?" name="check5" defaultChecked={true} isValid={true} />;
```

- Disabled

```tsx
import { CheckboxField } from '@zopauk/react-components';

<CheckboxField label="Do you love rap?" name="check6" disabled={true} />;
```

- Disabled, valid and checked

```tsx
import { CheckboxField } from '@zopauk/react-components';

<CheckboxField label="Do you love rap?" name="check7" isValid={true} defaultChecked={true} disabled={true} />;
```
