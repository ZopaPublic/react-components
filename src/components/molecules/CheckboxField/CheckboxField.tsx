import React, { forwardRef } from 'react';
import styled, { keyframes, css } from 'styled-components';
import { colors, typography } from '../../../constants';
import tealCheckMark from '../../../content/images/teal-check-mark.svg';
import greenCheckMark from '../../../content/images/green-check-mark.svg';
import unbrandedTick from '../../../content/images/unbranded-tick.svg';
import ErrorMessage from '../../atoms/ErrorMessage/ErrorMessage';
import InputLabel from '../../atoms/InputLabel/InputLabel';
import SizedContainer from '../../layout/SizedContainer/SizedContainer';
import { getBorderColorByStatus } from '../../../helpers/utils';
import { FieldProps, InputProps, GroupingControlsProps } from '../../types';
import { AppTheme, AppThemeProps, useThemeContext } from '../../styles/Theme';

export interface CheckboxFieldProps
  extends FieldProps,
    GroupingControlsProps,
    Omit<InputProps, 'startIcon' | 'endIcon'> {}

const getCheckedColor = ({ disabled, isValid, theme }: Pick<InputProps, 'disabled' | 'isValid'> & AppThemeProps) => {
  if (isValid) {
    return theme.input.borderColorByStatus.valid;
  }
  if (disabled) {
    return theme.input.borderColorByStatus.disabled;
  }
  return theme.input.checkBox.defaultColor;
};

const getTickIcon = ({ isValid, theme }: Pick<InputProps, 'disabled' | 'isValid'> & AppThemeProps) => {
  if (theme.input.checkBox.customIcon) {
    return unbrandedTick;
  }
  if (isValid) {
    return greenCheckMark;
  }
  return tealCheckMark;
};

const zoomOut = keyframes`
  from {
    transform: scale(0.4);
  }
  to {
    transform: scale(1);
  }
`;

const Label = styled(InputLabel)<Pick<InputProps, 'disabled' | 'hasError' | 'isValid'> & { theme: AppTheme }>`
  width: auto;
  display: flex;
  text-align: left;
  font-weight: ${typography.weights.regular};
  font-size: ${typography.sizes.text.body};
  padding: 14px 16px;
  border: 1px solid ${getBorderColorByStatus};
  transition-property: border, box-shadow;
  transition: 0.2s ease-in-out;
  border-radius: ${({ theme }) => theme.input.checkBox.label.borderRadius};
  line-height: 1.4;
  color: ${colors.greyDarkest};
  position: relative;
  user-select: none;
  margin-bottom: 0;

  &:before {
    content: '';
    flex-shrink: 0;
    background-color: ${colors.white};
    border-radius: ${({ theme }) => theme.input.checkBox.borderRadius};
    height: 20px;
    width: 20px;
    margin-right: 8px;
    transition-property: border, box-shadow;
    transition: 0.2s ease-in-out;
    box-shadow: 0 0 4px 0 transparent;
    border: 1px ${getBorderColorByStatus} solid;
    display: block;
  }
  &:after {
    content: '';
    background-size: contain;
    background: transparent none no-repeat center;
    height: 12px;
    width: 12px;
    position: absolute;
    left: 20px;
    top: 18px;
  }
  &:hover {
    cursor: pointer;
  }
`;

const Input = styled.input<InputProps & GroupingControlsProps>`
  left: -100%;
  opacity: 0;
  z-index: -1;
  position: absolute;

  &:hover:not(:disabled) + label,
  &:focus + label {
    border-color: ${({ theme }) => theme.input.checkBox.defaultColor};
    box-shadow: ${({ theme }) => theme.input.hover.boxShadow};
    &:before {
      border-color: ${({ theme }) => theme.input.checkBox.defaultColor};
      box-shadow: ${({ theme }) => theme.input.hover.boxShadow};
    }
  }
  &:hover:checked:not(:disabled) + label,
  &:focus:checked + label {
    &:after {
      background-image: ${({ theme }) =>
        theme.input.checkBox.customIcon ? `url(${unbrandedTick})` : `url(${tealCheckMark})`};
    }
  }
  &:disabled + label {
    cursor: not-allowed;
    color: ${colors.grey};
    background-color: ${colors.greyLightest};
  }
  &:disabled:not(:checked) + label {
    border-color: ${colors.greyLight};
    &:before {
      border-color: ${colors.greyLight};
    }
  }
  &:checked + label {
    border-color: ${getCheckedColor};
    background-color: ${({ theme }) => theme.input.checkBox.label.backgroundColor};
    box-shadow: none;
    &:before {
      border-color: ${getCheckedColor};
      box-shadow: none;
    }
    &:after {
      background-size: contain;
      background-image: url(${getTickIcon});
      animation: ${zoomOut} 180ms ease-in-out;
    }
  }
  ${({ hideControl }) =>
    hideControl &&
    css`
      & + label {
        &:before,
        &:after {
          display: none;
        }
      }
    `}
`;

const CheckboxField = forwardRef<HTMLInputElement, CheckboxFieldProps>((props, ref) => {
  const { label, errorMessage, className, inputSize, name, hasError, isValid, hideControl, ...rest } = props;

  const theme = useThemeContext();

  return (
    <>
      <SizedContainer size={inputSize} className={className}>
        <Input
          ref={ref}
          id={`checkbox-id-${name}`}
          type="checkbox"
          hasError={hasError}
          isValid={isValid}
          name={name}
          hideControl={hideControl}
          {...rest}
          theme={theme}
        />
        <Label htmlFor={`checkbox-id-${name}`} hasError={hasError} isValid={isValid} theme={theme}>
          {label}
        </Label>
      </SizedContainer>
      {errorMessage && <ErrorMessage className="mt-1">{errorMessage}</ErrorMessage>}
    </>
  );
});

export default CheckboxField;
