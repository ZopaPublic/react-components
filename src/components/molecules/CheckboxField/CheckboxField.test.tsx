import React from 'react';
import { render } from '@testing-library/react';
import axe from '../../../../axe-helper';
import CheckboxField from './CheckboxField';

describe('<CheckboxField />', () => {
  it('renders the component with props with no a11y violations', async () => {
    const { container } = render(<CheckboxField label={'hello'} name="test1" />);
    const results = await axe(container.innerHTML);

    expect(container.firstChild).toMatchSnapshot();
    expect(results).toHaveNoViolations();
  });

  it('renders the component with no control icon and no a11y violations', async () => {
    const { container } = render(<CheckboxField label={'hello'} name="test1" hideControl />);
    const results = await axe(container.innerHTML);

    expect(container.firstChild).toMatchSnapshot();
    expect(results).toHaveNoViolations();
  });
});
