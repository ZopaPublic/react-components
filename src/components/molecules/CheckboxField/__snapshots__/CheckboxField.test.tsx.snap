// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxField /> renders the component with no control icon and no a11y violations 1`] = `
.c2 {
  display: block;
  margin: 0 0 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c0 {
  max-width: 100%;
  width: 100%;
}

.c3 {
  width: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
  font-weight: 400;
  font-size: 16px;
  padding: 14px 16px;
  border: 1px solid #818F9B;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  border-radius: 8px;
  line-height: 1.4;
  color: #2C3236;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin-bottom: 0;
}

.c3:before {
  content: '';
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #FFFFFF;
  border-radius: 4px;
  height: 20px;
  width: 20px;
  margin-right: 8px;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  box-shadow: 0 0 4px 0 transparent;
  border: 1px #818F9B solid;
  display: block;
}

.c3:after {
  content: '';
  background-size: contain;
  background: transparent none no-repeat center;
  height: 12px;
  width: 12px;
  position: absolute;
  left: 20px;
  top: 18px;
}

.c3:hover {
  cursor: pointer;
}

.c1 {
  left: -100%;
  opacity: 0;
  z-index: -1;
  position: absolute;
}

.c1:hover:not(:disabled) + label,
.c1:focus + label {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c1:hover:not(:disabled) + label:before,
.c1:focus + label:before {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c1:hover:checked:not(:disabled) + label:after,
.c1:focus:checked + label:after {
  background-image: url(teal-check-mark.svg);
}

.c1:disabled + label {
  cursor: not-allowed;
  color: #818F9B;
  background-color: #F7F7F7;
}

.c1:disabled:not(:checked) + label {
  border-color: #D4D7D9;
}

.c1:disabled:not(:checked) + label:before {
  border-color: #D4D7D9;
}

.c1:checked + label {
  border-color: #009E8E;
  background-color: #ECFCF9;
  box-shadow: none;
}

.c1:checked + label:before {
  border-color: #009E8E;
  box-shadow: none;
}

.c1:checked + label:after {
  background-size: contain;
  background-image: url(teal-check-mark.svg);
  -webkit-animation: guYtTV 180ms ease-in-out;
  animation: guYtTV 180ms ease-in-out;
}

.c1 + label:before,
.c1 + label:after {
  display: none;
}

<div
  class="c0"
>
  <input
    class="c1"
    id="checkbox-id-test1"
    name="test1"
    type="checkbox"
  />
  <label
    class="c2 c3"
    for="checkbox-id-test1"
  >
    hello
  </label>
</div>
`;

exports[`<CheckboxField /> renders the component with props with no a11y violations 1`] = `
.c2 {
  display: block;
  margin: 0 0 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c0 {
  max-width: 100%;
  width: 100%;
}

.c3 {
  width: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
  font-weight: 400;
  font-size: 16px;
  padding: 14px 16px;
  border: 1px solid #818F9B;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  border-radius: 8px;
  line-height: 1.4;
  color: #2C3236;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin-bottom: 0;
}

.c3:before {
  content: '';
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #FFFFFF;
  border-radius: 4px;
  height: 20px;
  width: 20px;
  margin-right: 8px;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  box-shadow: 0 0 4px 0 transparent;
  border: 1px #818F9B solid;
  display: block;
}

.c3:after {
  content: '';
  background-size: contain;
  background: transparent none no-repeat center;
  height: 12px;
  width: 12px;
  position: absolute;
  left: 20px;
  top: 18px;
}

.c3:hover {
  cursor: pointer;
}

.c1 {
  left: -100%;
  opacity: 0;
  z-index: -1;
  position: absolute;
}

.c1:hover:not(:disabled) + label,
.c1:focus + label {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c1:hover:not(:disabled) + label:before,
.c1:focus + label:before {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c1:hover:checked:not(:disabled) + label:after,
.c1:focus:checked + label:after {
  background-image: url(teal-check-mark.svg);
}

.c1:disabled + label {
  cursor: not-allowed;
  color: #818F9B;
  background-color: #F7F7F7;
}

.c1:disabled:not(:checked) + label {
  border-color: #D4D7D9;
}

.c1:disabled:not(:checked) + label:before {
  border-color: #D4D7D9;
}

.c1:checked + label {
  border-color: #009E8E;
  background-color: #ECFCF9;
  box-shadow: none;
}

.c1:checked + label:before {
  border-color: #009E8E;
  box-shadow: none;
}

.c1:checked + label:after {
  background-size: contain;
  background-image: url(teal-check-mark.svg);
  -webkit-animation: guYtTV 180ms ease-in-out;
  animation: guYtTV 180ms ease-in-out;
}

<div
  class="c0"
>
  <input
    class="c1"
    id="checkbox-id-test1"
    name="test1"
    type="checkbox"
  />
  <label
    class="c2 c3"
    for="checkbox-id-test1"
  >
    hello
  </label>
</div>
`;
